demo/realtime-img2img/img2img.py

```diff
@@ -0,0 +1,105 @@
+default_negative_prompt = "black and white, blurry, low resolution, pixelated,  pixel art, low quality, low fidelity"
+
+page_content = """<h1 class="text-3xl font-bold">StreamDiffusion</h1>
+<h3 class="text-xl font-bold">Image-to-Image SD-Turbo</h3>
+<p class="text-sm">
+    This demo showcases
+    <a
+    href="https://github.com/cumulo-autumn/StreamDiffusion"
+    target="_blank"
+    class="text-blue-500 underline hover:no-underline">StreamDiffusion
+</a>
+Image to Image pipeline using
+    <a
+    href="https://huggingface.co/stabilityai/sd-turbo"
+    target="_blank"
+    class="text-blue-500 underline hover:no-underline">SD-Turbo</a
+    > with a MJPEG stream server.
+</p>
+"""
+
+
+class Pipeline:
+    class Info(BaseModel):
+        name: str = "StreamDiffusion img2img"
+        input_mode: str = "image"
+        page_content: str = page_content
+
+    class InputParams(BaseModel):
+        prompt: str = Field(
+            default_prompt,
+            title="Prompt",
+            field="textarea",
+            id="prompt",
+        )
+        # negative_prompt: str = Field(
+        #     default_negative_prompt,
+        #     title="Negative Prompt",
+        #     field="textarea",
+        #     id="negative_prompt",
+        # )
+        width: int = Field(
+            512, min=2, max=15, title="Width", disabled=True, hide=True, id="width"
+        )
+        height: int = Field(
+            512, min=2, max=15, title="Height", disabled=True, hide=True, id="height"
+        )
+
+    def __init__(self, args: Args, device: torch.device, torch_dtype: torch.dtype):
+        params = self.InputParams()
+        self.stream = StreamDiffusionWrapper(
+            model_id_or_path=base_model,
+            use_tiny_vae=args.taesd,
+            device=device,
+            dtype=torch_dtype,
+            t_index_list=[35, 45],
+            frame_buffer_size=1,
+            width=params.width,
+            height=params.height,
+            use_lcm_lora=False,
+            output_type="pil",
+            warmup=10,
+            vae_id=None,
+            acceleration=args.acceleration,
+            mode="img2img",
+            use_denoising_batch=True,
+            cfg_type="none",
+            use_safety_checker=args.safety_checker,
+            enable_similar_image_filter=True,
+            similar_image_filter_threshold=0.98,
+            engine_dir=args.engine_dir,
+        )
+
+        self.last_prompt = default_prompt
+        self.stream.prepare(
+            prompt=default_prompt,
+            negative_prompt=default_negative_prompt,
+            num_inference_steps=50,
+            guidance_scale=1.2,
+        )
+
+    def predict(self, params: "Pipeline.InputParams") -> Image.Image:
+        image_tensor = self.stream.preprocess_image(params.image)
+        for _ in range(self.stream.batch_size - 1):
+            self.stream(image=image_tensor, prompt=params.prompt)
```

Comment:
- is repeating inference multiple times needed here? I don't believe so, but if it is I would be keen to know why
- in the `"img2img"` mode it appears `stream.preprocess_image()` is already called, so I believe that is also unnecessary here
- it appears `img2img()` in `utils/wrapper.py` doesn't accept a `prompt` argument so I'm curious to know how the call to `stream()` works correctly here - presumably I've overlooked something because I see the prompt change clearly taking effect in the demo gifs

Reply:
I copy the batches call from here, it might be necessary, I'm no sure, wdyt @cumulo-autumn, same image tensor called `batch_size - 1` times 

https://github.com/cumulo-autumn/StreamDiffusion/blob/62596fd39b86a3088e28b2f2c58caa36bc069c50/examples/img2img/single.py#L97-L103

do you mean the `self.stream(image=image_tensor` can be a PIL image? my guess is I need to transform the `params.image` to a tensor every single predict `image_tensor = self.stream.preprocess_image(params.image)`

Reply:
Lines 100 and 101 of img2img.py are unnecessary.

We still need to do inference multiple times in img2img/single.py demo because we need to push out the initial latent tensor, which is initialized with torch.zeros(). 
```py
for _ in range(stream.batch_size - 1): 
	     stream(image=image_tensor) 
```
Since we are using Stream Batch, which batching the denoising steps, we need to push out initial n-1 images made from torch.zeros tensor first and then get the actual img2img result after that, where n is the number of denoising steps (n = len(t_index_list) ).

The concepts of Stream Batch are shown in Figure 2. and Figure 5. of our paper. https://arxiv.org/abs/2312.12491

If the Stream Batch is filled with zero tensors at the start, you need to call `stream()` multiple times to discard them (in this case, it's not necessary to use the same image multiple times). On the other hand, once the batch is filled with image tensors in the process of denoising, there's no need to discard those images in the middle of the StreamBatch. Instead, you can simply call stream() once within the while loop.

Reply:
So the short answers are as follows,
realtime-img2img/img2img.py ... repeatedly calling `stream()` is unnecessary.

img2img/single.py ... repeatedly calling `stream()` is necesary.

Reply:
perfect, removed unnecessary calls